plugin/storage/es/spanstore/reader_test.go

```diff
@@ -0,0 +1,420 @@
+		r.client.On("IndexExists", mock.AnythingOfType("string")).Return(existsService)
+
+		searchService := &mocks.SearchService{}
+		searchService.On("Type", stringMatcher(serviceType)).Return(searchService)
+		searchService.On("Size", mock.MatchedBy(func(i int) bool {
+			return i == 0
+		})).Return(searchService)
+		searchService.On("Aggregation",
+			stringMatcher("distinct_services"),
+			mock.AnythingOfType("*elastic.TermsAggregation")).Return(searchService)
+		r.client.On("Search", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string")).Return(searchService)
+
+		searchService.On("Do", mock.AnythingOfType("*context.emptyCtx")).
+			Return(testCase.searchResult, testCase.searchError)
+
+		actual, err := r.reader.GetServices()
+		require.Error(t, err, "Search failure")
+		assert.Nil(t, actual)
+	})
+}
+
+func TestSpanReader_GetServicesAggregationError(t *testing.T) {
+	aggregations := make(map[string]*json.RawMessage)
+
+	rawMessage := []byte(`{"buckets": [{bad json]}asdf`)
+	aggregations["distinct_services"] = (*json.RawMessage)(&rawMessage)
+
+	testCase := struct {
+		searchResult *elastic.SearchResult
+		searchError  error
+	}{
+		searchResult: &elastic.SearchResult{
+			Aggregations: elastic.Aggregations(aggregations),
+		},
+	}
+	withSpanReader(func(r *spanReaderTest) {
+		existsService := &mocks.IndicesExistsService{}
+		existsService.On("Do", mock.AnythingOfType("*context.emptyCtx")).Return(true, nil)
+		r.client.On("IndexExists", mock.AnythingOfType("string")).Return(existsService)
+
+		searchService := &mocks.SearchService{}
+		searchService.On("Type", stringMatcher(serviceType)).Return(searchService)
+		searchService.On("Size", mock.MatchedBy(func(i int) bool {
+			return i == 0
+		})).Return(searchService)
+		searchService.On("Aggregation",
+			stringMatcher("distinct_services"),
+			mock.AnythingOfType("*elastic.TermsAggregation")).Return(searchService)
+		r.client.On("Search", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string")).Return(searchService)
+
+		searchService.On("Do", mock.AnythingOfType("*context.emptyCtx")).
+			Return(testCase.searchResult, testCase.searchError)
+
+		actual, err := r.reader.GetServices()
+		require.Error(t, err, "Could not find aggregation of services")
+		assert.Nil(t, actual)
+	})
+}
+
+func TestSpanReader_getServicesAggregation(t *testing.T) {
+	// "terms":{
+	//    "size":3000,
+	//    "field":"serviceName"
+	// }
+	withSpanReader(func(r *spanReaderTest) {
+		serviceAggregation := r.reader.getServicesAggregation()
+		actual, err := serviceAggregation.Source()
+		require.NoError(t, err)
+		expected := make(map[string]interface{})
+		terms := make(map[string]interface{})
+		expected["terms"] = terms
+		terms["size"] = 3000
+		terms["field"] = "serviceName"
+
+		assert.EqualValues(t, expected, actual)
+	})
+}
+
+func TestSpanReader_GetOperations(t *testing.T) {
```

Comment:
gotcha. check the most recent commit, i kept the two tests sort of separate since they are separate functions, but pulled most of it out into separate helper fns so that i dont have the same code lying around everywhere